pkg/leeway/workspace.go

```diff
@@ -36,13 +44,79 @@ type Workspace struct {
 	ignores []string
 }
 
-// ShouldIngoreComponent returns true if a file should be ignored for a component listing
-func (ws *Workspace) ShouldIngoreComponent(path string) bool {
-	return ws.ShouldIngoreSource(path)
+// EnvironmentManifest is a collection of environment manifest entries
+type EnvironmentManifest []EnvironmentManifestEntry
+
+// Write writes the manifest to the writer
+func (mf EnvironmentManifest) Write(out io.Writer) error {
+	for _, e := range mf {
+		_, err := fmt.Fprintf(out, "%s: %s\n", e.Name, e.Value)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Hash produces the hash of this manifest
+func (mf EnvironmentManifest) Hash() (string, error) {
+	key, err := hex.DecodeString(contentHashKey)
+	if err != nil {
+		return "", err
+	}
+
+	hash, err := highwayhash.New(key)
+	if err != nil {
+		return "", err
+	}
+
+	err = mf.Write(hash)
+	if err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(hash.Sum(nil)), nil
+}
+
+// EnvironmentManifestEntry represents an entry in the environment manifest
+type EnvironmentManifestEntry struct {
+	Name    string   `yaml:"name"`
+	Command []string `yaml:"command"`
+
+	Value   string `yaml:"-"`
+	Builtin bool   `yaml:"-"`
+}
+
+const (
+	builtinEnvManifestGOOS   = "goos"
+	builtinEnvManifestGOARCH = "goarch"
+)
+
+var defaultEnvManifestEntries = map[PackageType]EnvironmentManifest{
+	"": []EnvironmentManifestEntry{
+		{Name: "os", Command: []string{builtinEnvManifestGOOS}, Builtin: true},
+		{Name: "arch", Command: []string{builtinEnvManifestGOARCH}, Builtin: true},
+	},
+	GenericPackage: []EnvironmentManifestEntry{},
+	DockerPackage: []EnvironmentManifestEntry{
+		{Name: "docker", Command: []string{"docker", "version", "--format", "{{.Client.Version}}"}},
```

Comment:
If the build depends on the version of the docker client then it'd make sense to also add `{{ .Server.Version }}` as that should have more impact on the build output.

Reply:
Changed